File: demo/node/src/handlers.ts
import { Connection as AmqpConnection } from "@gomessaging/amqp";
import { Publisher as AmqpPublisher } from "@gomessaging/amqp";
import { Connection as NatsConnection } from "@gomessaging/nats";
import { Publisher as NatsPublisher } from "@gomessaging/nats";
import type { ConsumableEvent } from "@gomessaging/spec";
import type {
  OrderCreated,
  PingMessage,
  EmailRequest,
  EmailResponse,
} from "./messages.js";
import type { SSEEvent } from "./web.js";
import { randomUUID } from "node:crypto";

export interface AmqpSetup {
  conn: AmqpConnection;
  publisher: AmqpPublisher;
}

export interface NatsSetup {
  conn: NatsConnection;
  publisher: NatsPublisher;
}

export function setupAmqp(
  amqpUrl: string,
  broadcast: (event: SSEEvent) => void,
): AmqpSetup {
  const conn = new AmqpConnection({
    url: amqpUrl,
    serviceName: "node-demo",
    onClose: (err) => {
      console.error("AMQP connection lost, shutting down:", err.message);
      process.exit(1);
    },
  });

  const publisher = conn.addEventPublisher();

  conn.addEventConsumer<OrderCreated>("Order.Created", async (event) => {
    console.log(
      `[amqp] received Order.Created from ${event.payload.source}: ${event.payload.orderId}`,
    );
    broadcast({
      type: "received",
      transport: "amqp",
      source: event.payload.source,
      routingKey: event.deliveryInfo.key,
      payload: event.payload,
      traceId: event.id,
    });
  });

  conn.addEventConsumer<PingMessage>("Ping", async (event) => {
    console.log(
      `[amqp] received Ping from ${event.payload.source}: ${event.payload.message}`,
    );
    broadcast({
      type: "received",
      transport: "amqp",
      source: event.payload.source,
      routingKey: event.deliveryInfo.key,
      payload: event.payload,
      traceId: event.id,
    });
  });

  conn.addServiceRequestConsumer<EmailRequest, EmailResponse>(
    "email.send",
    async (event: ConsumableEvent<EmailRequest>) => {
      console.log(
        `[amqp] handling email request to ${event.payload.to}`,
      );
      broadcast({
        type: "received",
        transport: "amqp",
        source: event.source,
        routingKey: event.deliveryInfo.key,
        payload: event.payload,
        traceId: event.id,
      });
      return {
        messageId: `msg-${randomUUID().substring(0, 8)}`,
        status: "sent",
      };
    },
  );

  return { conn, publisher };
}

export function setupNats(
  natsUrl: string,
  broadcast: (event: SSEEvent) => void,
): NatsSetup {
  const conn = new NatsConnection({ url: natsUrl, serviceName: "node-demo" });

  const publisher = conn.addEventPublisher();

  conn.addEventConsumer<OrderCreated>("Order.Created", async (event) => {
    console.log(
      `[nats] received Order.Created from ${event.payload.source}: ${event.payload.orderId}`,
    );
    broadcast({
      type: "received",
      transport: "nats",
      source: event.payload.source,
      routingKey: event.deliveryInfo.key,
      payload: event.payload,
      traceId: event.id,
    });
  });

  conn.addEventConsumer<PingMessage>("Ping", async (event) => {
    console.log(
      `[nats] received Ping from ${event.payload.source}: ${event.payload.message}`,
    );
    broadcast({
      type: "received",
      transport: "nats",
      source: event.payload.source,
      routingKey: event.deliveryInfo.key,
      payload: event.payload,
      traceId: event.id,
    });
  });

  conn.addServiceRequestConsumer<EmailRequest, EmailResponse>(
    "email.send",
    async (event: ConsumableEvent<EmailRequest>) => {
      console.log(
        `[nats] handling email request to ${event.payload.to}`,
      );
      broadcast({
        type: "received",
        transport: "nats",
        source: event.source,
        routingKey: event.deliveryInfo.key,
        payload: event.payload,
        traceId: event.id,
      });
      return {
        messageId: `msg-${randomUUID().substring(0, 8)}`,
        status: "sent",
      };
    },
  );

  return { conn, publisher };
}

File: demo/go/handlers.go
package main

import (
	"context"
	"fmt"
	"log/slog"

	"github.com/google/uuid"
	"github.com/sparetimecoders/messaging/golang/amqp"
	"github.com/sparetimecoders/messaging/golang/nats"
	"github.com/sparetimecoders/messaging/specification/spec"
)

// amqpSetups returns the AMQP setup functions for the go-demo service.
func amqpSetups(pub *amqp.Publisher, broadcast func(SSEEvent)) []amqp.Setup {
	return []amqp.Setup{
		amqp.EventStreamPublisher(pub),
		amqp.EventStreamConsumer("Order.Created", orderHandler("amqp", broadcast)),
		amqp.EventStreamConsumer("Ping", pingHandler("amqp", broadcast)),
		amqp.RequestResponseHandler("email.send", emailHandler("amqp", broadcast)),
	}
}

// natsSetups returns the NATS setup functions for the go-demo service.
func natsSetups(pub *nats.Publisher, broadcast func(SSEEvent)) []nats.Setup {
	return []nats.Setup{
		nats.EventStreamPublisher(pub),
		nats.EventStreamConsumer("Order.Created", orderHandler("nats", broadcast)),
		nats.EventStreamConsumer("Ping", pingHandler("nats", broadcast)),
		nats.RequestResponseHandler("email.send", emailHandler("nats", broadcast)),
	}
}

func orderHandler(transport string, broadcast func(SSEEvent)) spec.EventHandler[OrderCreated] {
	return func(ctx context.Context, event spec.ConsumableEvent[OrderCreated]) error {
		slog.Info("received Order.Created",
			"transport", transport,
			"source", event.Payload.Source,
			"orderId", event.Payload.OrderID,
		)
		broadcast(SSEEvent{
			Type:       "received",
			Transport:  transport,
			Source:     event.Payload.Source,
			RoutingKey: event.DeliveryInfo.Key,
			Payload:    event.Payload,
			TraceID:    event.Metadata.ID,
		})
		return nil
	}
}

func pingHandler(transport string, broadcast func(SSEEvent)) spec.EventHandler[PingMessage] {
	return func(ctx context.Context, event spec.ConsumableEvent[PingMessage]) error {
		slog.Info("received Ping",
			"transport", transport,
			"source", event.Payload.Source,
			"message", event.Payload.Message,
		)
		broadcast(SSEEvent{
			Type:       "received",
			Transport:  transport,
			Source:     event.Payload.Source,
			RoutingKey: event.DeliveryInfo.Key,
			Payload:    event.Payload,
			TraceID:    event.Metadata.ID,
		})
		return nil
	}
}

func emailHandler(transport string, broadcast func(SSEEvent)) spec.RequestResponseEventHandler[EmailRequest, EmailResponse] {
	return func(ctx context.Context, event spec.ConsumableEvent[EmailRequest]) (EmailResponse, error) {
		slog.Info("handling email request",
			"transport", transport,
			"to", event.Payload.To,
			"subject", event.Payload.Subject,
		)
		broadcast(SSEEvent{
			Type:       "received",
			Transport:  transport,
			Source:     event.Source,
			RoutingKey: event.DeliveryInfo.Key,
			Payload:    event.Payload,
			TraceID:    event.Metadata.ID,
		})
		resp := EmailResponse{
			MessageID: fmt.Sprintf("msg-%s", uuid.New().String()[:8]),
			Status:    "sent",
		}
		return resp, nil
	}
}
